videostore/rawsegmenter.go

```diff
@@ -0,0 +1,287 @@
+package videostore
+
+/*
+#include <libavfilter/avfilter.h>
+#include <libavutil/avutil.h>
+#include <libavformat/avformat.h>
+#include <libavutil/dict.h>
+#include <libavutil/error.h>
+#include <libavcodec/avcodec.h>
+*/
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"unsafe"
+
+	"github.com/bluenviron/mediacommon/pkg/codecs/h264"
+	"go.viam.com/rdk/logging"
+)
+
+type rawSegmenter struct {
+	logger         logging.Logger
+	storagePath    string
+	segmentSeconds int
+	outCtxMu       sync.Mutex
+	outCtx         *C.AVFormatContext
+}
+
+func newRawSegmenter(
+	logger logging.Logger,
+	storagePath string,
+	segmentSeconds int,
+) (*rawSegmenter, error) {
+	s := &rawSegmenter{
+		logger:         logger,
+		storagePath:    storagePath,
+		segmentSeconds: segmentSeconds,
+	}
+	err := createDir(s.storagePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (rs *rawSegmenter) initH264(sps, pps []byte) error {
+	if len(sps) == 0 || len(pps) == 0 {
+		return errors.New("both sps & pps must not be empty")
+	}
+
+	var hsps h264.SPS
+	if err := hsps.Unmarshal(sps); err != nil {
+		return err
+	}
+	rs.outCtxMu.Lock()
+	defer rs.outCtxMu.Unlock()
+	if rs.outCtx != nil {
+		return errors.New("*rawSegmenter init called more than once")
+	}
+	// Allocate output context for segmenter. The "segment" format is a special format
+	// that allows for segmenting output files. The output pattern is a strftime pattern
+	// that specifies the output file name. The pattern is set to the current time.
+	outputPatternCStr := C.CString(rs.storagePath + "/" + outputPattern)
+	defer C.free(unsafe.Pointer(outputPatternCStr))
+	formatName := C.CString("segment")
+	defer C.free(unsafe.Pointer(formatName))
+	var fmtCtx *C.AVFormatContext
+	ret := C.avformat_alloc_output_context2(&fmtCtx, nil, formatName, outputPatternCStr)
+	if ret < 0 {
+		return fmt.Errorf("failed to allocate output context: %s", ffmpegError(ret))
+	}
+	rs.outCtx = fmtCtx
+
+	// Create new stream for the output context.
+	stream := C.avformat_new_stream(fmtCtx, nil)
+	if stream == nil {
+		return errors.New("failed to allocate stream")
+	}
+	stream.id = C.int(fmtCtx.nb_streams) - 1
+	// stream.time_base.num = 1
+	// stream.time_base.den = 1
+
+	// Set the codec parameters for the output stream.
+	codec := C.avcodec_find_encoder(C.AV_CODEC_ID_H264)
+	if codec == nil {
+		return errors.New("failed to find codec")
+	}
+
+	codecCtx := C.avcodec_alloc_context3(codec)
+	if codecCtx == nil {
+		return errors.New("failed to allocate codec context")
+	}
+
+	// TODO(seanp); Size and pixel format is hardcoded for now. Need to make this dynamic.
+	// Maybe we can get this from ONVIF or the RTSP stream SDP.
+	codecCtx.width = C.int(hsps.Width())
+	codecCtx.height = C.int(hsps.Height())
+	// codecCtx.pix_fmt = C.AV_PIX_FMT_YUV420P
+	// codecCtx.time_base.num = 1
```

Comment:
Can we at least add a comment here that we are using the default time base? Might be better to be more explicit here and pass in the desired time base.